tensorflow/go/op/gradients.go

```diff
@@ -0,0 +1,36 @@
+/*
+Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package op
+
+import tf "github.com/tensorflow/tensorflow/tensorflow/go"
+
+// Gradients adds gradients computation ops to the graph according to scope.
+//
+// Arguments:
+//  y: output of the function to derive
+//  x: inputs of the function for which partial derivatives are computed
+//  dx: if not null, the partial derivatives of some loss function L w.r.t. y
+//
+//  return the partial derivatives
+func Gradients(scope *Scope, y []tf.Output, x []tf.Output, dx ...tf.Output) (output []tf.Output) {
+	var err error
+	if output, err = scope.graph.AddGradients(scope.opName(scope.uniqueName("Gradients")), y, x, dx); err != nil {
```

Comment:
Hmm...isn't this missing other elements of the `scope`? For example, control dependencies and device annotations (see https://github.com/tensorflow/tensorflow/blob/e45c90f0e4d17ac22048a73f1e81bd9c7a7a5145/tensorflow/go/op/scope.go#L85)

I'm okay if you'd want to be conservative and have this fall (by setting an error in `scope`) when `scope` had control dependencies or devices specified (that aren't handled here). Or better yet, handle them :)

Reply:
Sorry for the delay. I'm not sure what you are meaning about "missing other elements of the `scope`"...

If you are talking about `ControlDependencies` or `Device`, I think `WithControlDependencies` and `WithDevice` is already enough if who want to set them.

Of course we can also add it to `Gradient` and set it in `Gradient`,but the parameters will be more complex. 
Because the `dx` is already a optional parameters and it's a `[]Outputs`. So the declaration will be like below:
```
func Gradients(scope *Scope, y []tf.Output, x []tf.Output, option ...map[string]interface{}) (output []tf.Output)
```
And the call will be like:
```
sub := s.SubScope("sub")
grads1 := Gradients(sub, []tf.Output{y2}, {'dx':[]tf.Output{x1, x2}, 'device': '...', ''})
```
The implementation will be like:

Reply:
RE: "other elements of "scope", I believe the control dependencies and devices are _not_ being handled in the call to `Gradients(scope, ...)`.  

All the generated ops in the `op` package (like `Abs`, `Add` etc., see [`wrappers.go`](https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/go/op/wrappers.go)) invoke [`Scope.AddOperation`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/go/op/scope.go#L75), which in turns invokes [`Graph.AddOperation`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/go/graph.go#L191) which adds the control inputs and the device annotation via calls to `TF_AddControlInput` and `TF_SetDevice` etc.

However, the implementation of `Gradients` in this PR will not invoke those functions and thus will not respect any control dependencies or device annotations set in the `Scope` object.

Does that make sense? You could probably most easily see this by inspecting the `GraphDef` protocol buffer written out by `Graph.WriteTo()`.

I realize that fixing this would require some complexity (for example, the device annotation would probably have to be plumbed through the C API of `TF_AddGradients`, and the control dependencies may require wrapping the inputs in an `Identity` operation or something). So, for this PR, I'm okay if we don't handle them but instead have `Gradients()` fail if it sees non-empty control dependencies or a device in the `Scope`. Does that make sense?

Reply:
Thanks for explanation but I am sorry that I still have some confusion.

According the implementation and test case in C/C++ and Java. `AddGradient` is not like with other op of scope.

For example, we can see test case in C++, [TEST_F(GradientsTest, TwoMatMuls_Chained)](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/gradients_test.cc#L139-L170).
In this case, we can see the common op like `Const`/`MatMul` and gradient op `AddSymbolicGradients`. 
In [here](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/c/c_api_test.cc#L1695-L1710), These common op is implemented with `TF_NewOperation` and `TF_FinishOperation`, so they can invoke `TF_AddInput` or `TF_AddControlInput` or `TF_SetDevice`(All these functions need a `TF_OperationDescription` returned by `TF_NewOperation`.
But [`AddSymbolicGradients`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/gradients.cc#L557-L565) doesn't invoke `TF_NewOperation`, so it can't invoke `TF_AddControlInput` or `TF_SetDevice` because it don't have a `TF_OperationDescription`.

Our implementation in Go actually needs to invode C implementation. So I'm confused how to invoke `TF_AddControlInput` or `TF_SetDevice` when the c implementation doesn't invoke these function...
If we just use [`WithControlDependencies`](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/go/op/scope.go#L115-L133), it will be effective until next "common op" is used.

Reply:
In fact, [the implementation of C](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/c/c_api.cc#L2441-L2447) uses [the api about scope in C++](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/gradients.h#L30-L35).

And the scope of C++ also have [device](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/scope.h#L132-L134) and [controlDepandencies](https://github.com/tensorflow/tensorflow/blob/94ab1a9e82fb7630b1aae44984d73968508dc917/tensorflow/cc/framework/scope.h#L119-L126)
But the implementation of  gradient don't handle them.

I'm not sure if it is right to implement it with different logic in C++(when we finally call the implementation in C++).

Of course, if we still think it it more important to be consistent with other op in go or there are other details that I didn't find, I'm also willing to just lead to an error in the code now.
@asimshankar

Reply:
@asimshankar Can you take a look again?